services/transactionpool/forward_transactions.go

```diff
@@ -0,0 +1,151 @@
+package transactionpool
+
+import (
+	"context"
+	"github.com/orbs-network/orbs-network-go/crypto/digest"
+	"github.com/orbs-network/orbs-network-go/crypto/signature"
+	"github.com/orbs-network/orbs-network-go/instrumentation/log"
+	"github.com/orbs-network/orbs-network-go/synchronization"
+	"github.com/orbs-network/orbs-spec/types/go/primitives"
+	"github.com/orbs-network/orbs-spec/types/go/protocol"
+	"github.com/orbs-network/orbs-spec/types/go/protocol/gossipmessages"
+	"github.com/orbs-network/orbs-spec/types/go/services/gossiptopics"
+	"github.com/orbs-network/orbs-spec/types/go/services/handlers"
+	"github.com/pkg/errors"
+	"sync"
+	"time"
+)
+
+type TransactionForwarderConfig interface {
+	NodePublicKey() primitives.Ed25519PublicKey
+	NodePrivateKey() primitives.Ed25519PrivateKey
+	TransactionPoolPropagationBatchSize() uint16
+	TransactionPoolPropagationBatchingTimeout() time.Duration
+}
+
+func (s *service) RegisterTransactionResultsHandler(handler handlers.TransactionResultsHandler) {
+	s.transactionResultsHandlers = append(s.transactionResultsHandlers, handler)
+}
+
+func (s *service) HandleForwardedTransactions(input *gossiptopics.ForwardedTransactionsInput) (*gossiptopics.EmptyOutput, error) {
+	sender := input.Message.Sender
+	oneBigHash, _ := HashTransactions(input.Message.SignedTransactions...)
+
+	if !signature.VerifyEd25519(sender.SenderPublicKey(), oneBigHash, sender.Signature()) {
+		return nil, errors.Errorf("invalid signature in relay message from sender %s", sender.SenderPublicKey())
+	}
+
+	for _, tx := range input.Message.SignedTransactions {
+		txHash := digest.CalcTxHash(tx.Transaction())
+		s.logger.Info("adding forwarded transaction to the pool", log.String("flow", "checkpoint"), log.Stringable("transaction", tx), log.Stringable("txHash", txHash))
+		if _, err := s.pendingPool.add(tx, sender.SenderPublicKey()); err != nil {
+			s.logger.Error("error adding forwarded transaction to pending pool", log.Error(err), log.Stringable("transaction", tx), log.Stringable("txHash", txHash))
+		}
+	}
+	return nil, nil
+}
+
```

Comment:
also about this .. the service does the handling even if it is forwarded ... i'd consider leaving it in the service or maybe joining the two funcs to a new file called tx_handlers.go ?

Reply:
We decided to move everything related to transaction forwarding in the same file.